src/vsr.zig

```diff
@@ -1917,3 +1918,52 @@ const ViewChangeHeadersArray = struct {
         headers.array.appendAssumeCapacity(Headers.dvc_blank(op));
     }
 };
+
+pub const Op = struct {
```

Comment:
```suggestion
pub const Checkpoint = struct {
```

Since all the method names in here include `checkpoint`, and since checkpoints or checkpoint calculations are what this is really about.

To riff on Brian Eno's "Honor your mistake as a hidden intention" (e.g. when @kprotty kept saying "groove" we just had fun and together decided to rename `Grove` to `Groove`), this one could be an instance of "Honor your repetition as a namespace suggestion". 😛

Reply:
They are all checkpoint functions right now, but this namespace will also shortly (as part of table sync) get `snapshot_from_op`, which doesn't have anything to do with checkpoints.

Reply:
I do still much prefer `vsr.Checkpoint` over `Op` to keep these checkpoint-related functions together, and would then rather put op-related helpers in another namespace.

Also, do we need `snapshot_from_op`, or can we do without it? The `snapshot` should be the `op` number for a simple mental model.

Reply:
Ops are snapshots. But there are different ways they can correspond. Perhaps `snapshot_from_commit` would be a better name.

```zig
    /// A table with TableInfo.snapshot_min=S was written during some commit with op<S.
    /// A block with snapshot_min=S is definitely readable at op=S.
    pub fn snapshot_from_commit(op: u64) u64 {
        return op + 1;
    }
```

Roughly, "block.snapshot=X" is "X is the first op that can read `block`.

> Also, do we need snapshot_from_op, or can we do without it? 

We could inline that, but I think the label describing what is happening is useful for readability, even though the implementation is trivial.

Reply:
To clarify if we are on the same page (I think we are)... the intended sense of `table.snapshot_min` should be "the minimum op number that can see the table", as your comment above the function has it.

However, I don't follow the `return op + 1` above. It should be an identity function rather.

Reply:
For table/reply sync, we will store `commit_unsynced_min` and `commit_unsynced_max` in the superblock's VSRState:

```
        /// See `commit_unsynced_max`.
        commit_unsynced_min: u64,

        /// When zero, all of the grid and replies are synced.
        /// (When zero, `commit_unsynced_min` is also zero.)
        ///
        /// When nonzero, we must repair grid-blocks/client-replies that would have been written
        /// during the commits between `commit_unsynced_min` and `commit_unsynced_max` (inclusive).
        /// (They were not written normally because we skipped past them via state sync.)
        commit_unsynced_max: u64,
```

Then as we iterate the tables to figure out which need to be synced, and which we already have:

```
            assert(self.superblock.working.vsr_state.commit_unsynced_max > 0);

            const commit_unsynced_min = self.superblock.working.vsr_state.commit_unsynced_min;
            const commit_unsynced_max = self.superblock.working.vsr_state.commit_unsynced_max;
            while (self.sync_tables.?.next(&self.state_machine.forest)) |table_info| {
                if (table_info.snapshot_min >= vsr.Op.snapshot_from_commit(commit_unsynced_min) and
                    table_info.snapshot_min <= vsr.Op.snapshot_from_commit(commit_unsynced_max))
                {
                    // (sync the table)
                } else {
                    // (table is already synced)
                }
```

We could apply the `+1` when we set `commit_unsynced_min`/`max` (instead of in `snapshot_from_commit`), but I think that results in a less intuitive mental model of what those ops mean.

Reply:
>to rename Grove to Groove)

:exploding_head: these are different words!

Reply:
Thanks for the explanation.

Stepping back from state sync, for snapshot usage in general (e.g. for read transactions), I don't think we should use a helper (or it should be an identity function), and the snapshot value should be the committed op at which the snapshot was acquired.

I would rather change the inclusive/exclusive semantics of the unsynced range, then change the identity function between op/snapshot.

Reply:
I guess we could go full OOP here?

```
pub const Checkpoint = struct {
    op: u64,
    trigger: u64,

    fn next(self: Checkpoint) Checkpoint {

    }

    fn prev(self: Checkpoint) Checkpoint {

    }
};
```

The call site would be `replica.checkpoint().next().trigger`

Reply:
> the snapshot value should be the committed op at which the snapshot was acquired.

I don't understand what you are suggesting. (We haven't actually implemented snapshot acquisition so maybe our mental models don't match up?)

Do you mean that, when we send a message to acquire a user snapshot, that the op on the reply message is our snapshot? (Which means that we would "backdate" the snapshots of writes of data that is currently in memory, I guess?)

> then change the identity function between op/snapshot.

These are all different ops:
1. when data arrives at a replica (the WAL + mutable table),
2. when data starts writing to disk (the first beat of a bar, but delayed by ~bar from arrival),
3. when data is definitely written to disk (the end of a bar, again delayed by ~bar from arrival) 

So wrt "the identity function between op/snapshot", which gets the identity function? Because it can be at most one of them :sweat_smile: 

(Right now it is 3, which is imo the most natural choice implementation-wise (which was my reason for choosing it).)

---

> I would rather change the inclusive/exclusive semantics of the unsynced range, then change the identity function between op/snapshot.

Fwiw, this is really the most natural semantics for the range, e.g. when determining which client replies to sync:

```zig
                    entry.header.op >= self.superblock.working.vsr_state.commit_unsynced_min and
                    entry.header.op <= self.superblock.working.vsr_state.commit_unsynced_max;
```

Reply:
> I don't understand what you are suggesting. (We haven't actually implemented snapshot acquisition so maybe our mental models don't match up?)

Thanks. Yes, I'm referring to how snapshots will be implemented. What "snapshot" will mean w.r.t. to the "op" that acquires it.

> So wrt "the identity function between op/snapshot", which gets the identity function?

I'm thinking not at the storage layer but at the state machine layer. For the op that acquires the snapshot, whatever the op can see (if it were executing say a read transaction) is what the snapshot should see.

By "identity function", I mean that if an op is to acquire a snapshot, then the value of the snapshot acquired is that op number.

How we implement this under the hood will require pulsing, to get the acquiring op onto a bar boundary.

Reply:
Ok, thanks for the clarification, I think I understand now!

---

Taking a step back: As the "user-visible" snapshot id, why not use (state machine) timestamps instead of op numbers?

- Op-numbers are currently "internal" to the vsr-replica/vsr-client world -- op-snapshots would expose them to users.
- Timestamps have more semantic meaning -- they are self-describing.
- We plan to have "scheduled future snapshots", which also play very well with this idea. You schedule the snapshot with a particular timestamp, and then if you query it before then it just says "scheduled but not queryable yet".
- We could map the snapshot-timestamps to snapshot-ops internally where we keep a record of persisted snapshots (in the superblock header iirc), so we can still use the ops to manage table visibility.

Reply:
That's a great idea.

As you say, physical timestamps, over logical op timestamps, also have the value of being auditable/relatable in the real world.